### ISSUE TITLE
[$1000] The emoji recommendation list closes on switching to next tab

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16576]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to staging dot on chrome
2. Go to a chat > Type in any emoji within colon. Example :dis and the recommendation list appears.
3. Switch to the next tab and come back to the application.
4. Result : See that the emoji recommendation list closes
5. Follow the above steps in slack and see that the emoji recommendation doesn’t close even when we switch to the next tab

## Expected Result:
The emoji recommendation list shouldn’t close on switching to the next tab

## Actual Result:
The emoji recommendation list closes on switching to the next tab

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.90-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/228036455-823e9754-ea41-44d6-b36d-5f1e6d31a0b0.mp4

https://user-images.githubusercontent.com/43996225/228036470-f93e8a4d-5bef-4cb0-9025-d7a7ee0d102c.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @avi-shek-jha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679907956819009

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010419d96d34cc6f01</li>
        <li>Upwork Job ID: 1646895022138916864</li>
        <li>Last Price Increase: 2023-05-15</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by Pujan92** [comment]
Maybe dupe of https://github.com/Expensify/App/issues/16364

**Comment by MelvinBot** [comment]
@michaelhaxhiu Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@michaelhaxhiu 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by michaelhaxhiu** [comment]
Thanks @Pujan92.

I'd like to cc @0xmiroslav - curious what you think about this bug? If we agree with your latest comment on https://github.com/Expensify/App/issues/16364#issuecomment-1493702177 then I think we would proceed with calling this a bug?

>So can we confirm that consistent behavior should be to show recommendation on both focus and selection change always on all platforms?

**Comment by michaelhaxhiu** [comment]
I can't decide if this is a feature or bug, basically?

**Comment by chiragsalian** [comment]
Discussed on [slack](https://expensify.slack.com/archives/C03U7DCU4/p1680803928476979) and it's decided as not a priority for the moment. Closing it out.

**Comment by michaelhaxhiu** [comment]
We had a change of heart after discussing further, let's re-open and prioritize because this is a valid inconsistency in the product (and thus - we should fix it).

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~010419d96d34cc6f01

**Comment by MelvinBot** [comment]
Current assignee @michaelhaxhiu is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcochavezf (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The emoji recommendation list automatically closes if a user switches to a different browser tab and returns back to the Expensify Chat tab.

### What is the root cause of that problem?
The emoji recommendation list is set to hide with a call to `this.resetSuggestedEmojis`  in the `onBlur` event handler for the composer on ReportActionCompose.

### What changes do you think we should make in order to solve the problem?
We can remove this reset method call, and check if the user manually changes focus away from the composer input.

Alternatively, we can add code to check the content of the composer when it regains focus and re-display the emoji suggestion based on the current input and the cursor position.

### What alternative solutions did you explore? (Optional)
None.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The emoji recommendation list closes on switching to the next tab

### What is the root cause of that problem?
In https://github.com/Expensify/App/blob/a5d171b6ccd6c2cbef00285c4d984796bf4b845d/src/pages/home/report/ReportActionCompose/index.js#L859 we're resetting the suggested emojis when the `Composer` is blurred, thus close the emoji recommendation list.

If we mirror the correct behavior of Slack, in the following 2 cases we should not close the emoji recommendation list when the composer is blurred:
1. When we switch to another browser tab
2. When we open another native app (aside from the browser)

### What changes do you think we should make in order to solve the problem?

In here https://github.com/Expensify/App/blob/a5d171b6ccd6c2cbef00285c4d984796bf4b845d/src/pages/home/report/ReportActionCompose/index.js#L859, we need to check that the window is currently visible and focused, if yes then we'll reset the emoji list since that means the user proactively blurs the input. If no, we should not reset the emoji list since the user only switches to another tab/app and will come back and focus on the input again.

We can use `Visibility.isVisible` here https://github.com/Expensify/App/blob/4607dcb69b6d5bb7607f652f8a78327c7e002183/src/libs/Visibility/index.js#L8 for this. However there's currently a bug in  `Visibility.isVisible` in web that causes it to return true even when `we open another native app`, that's being addressed here https://github.com/Expensify/App/issues/17242. That issue will be fixed by updating the `Visibility.isVisible` to `return document.visibilityState === 'visible' && document.hasFocus()` as per the [accepted proposal there](https://github.com/Expensify/App/issues/17242). 

That's exactly what we need so after that change is merged, we can just update https://github.com/Expensify/App/blob/a5d171b6ccd6c2cbef00285c4d984796bf4b845d/src/pages/home/report/ReportActionCompose/index.js#L859 to 
```
if (Visibility.isVisible()) {
    this.resetSuggestedEmojis();
}
```

### What alternative solutions did you explore? (Optional)
NA

**Comment by MelvinBot** [comment]
@marcochavezf @michaelhaxhiu @0xmiroslav this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by MelvinBot** [comment]
@marcochavezf, @michaelhaxhiu, @0xmiroslav Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by dukenv0307** [comment]
> Because composer is blurred, blue outline is removed but emoji suggestions still remaining.

@0xmiroslav Slack also works like that so I think that's fine. I think the user in that case is ready to return to the app to continue to select emojis so keeping the emoji suggestions is ok.

**Comment by MelvinBot** [comment]
@marcochavezf @michaelhaxhiu @0xmiroslav this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @0xmiroslav is eligible for the Internal assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @michaelhaxhiu is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @0xmiroslav is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @marcochavezf is eligible for the External assigner, not assigning anyone new.

**Comment by akinwale** [comment]
@0xmiroslav Have you had a chance to review my [proposal](https://github.com/Expensify/App/issues/16576#issuecomment-1508985896)?

**Comment by akinwale** [comment]
> > @0xmiroslav Have you had a chance to review my [proposal](https://github.com/Expensify/App/issues/16576#issuecomment-1508985896)?
> 
> @akinwale yes I checked. Your solution doesn't work for me.

Are you testing in Chrome or Safari? I can take another look and try to come up with a better solution.

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The emoji suggestions are hidden when we open a new tab

### What is the root cause of that problem?

In the onBlur of the composer we are reseting the emojis suggestions

### What changes do you think we should make in order to solve the problem?


We can subscribe to AppState change to know if the app is in the background or not:
`AppState.addEventListener('change', this._handleAppStateChange);`

Unfortunely the appState change happens after the onBlur so in the onBlur instead of calling resetSuggestedEmojis  we can:
- put this.state.shouldShowSuggestionMenu to false
- in 20ms put this.state.suggestedEmojis to [] if the app State is "active" because that means we are still in the app

In the new method _handleAppStateChange, if the app is now in the background and this.state.suggestedEmojis is not [] that means the suggested emojis were opened and we can put shouldShowSuggestionMenu to true so that when the user comes back to the app he will see the suggested emojis again.

That's the result I have:

https://user-images.githubusercontent.com/19537677/234995922-1df94149-171a-4a3e-8032-12da14e2316d.mp4

It's not clear in the video but when I click another app (for example NotPad) the suggested emojis are hidden, I didn't understand if it is what we want or not.

### What alternative solutions did you explore? (Optional)

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by michaelhaxhiu** [comment]
Checking in - how is this coming along so far @0xmiroslav? Do we have a proposal that's viable or need to go `internal`?

**Comment by marcochavezf** [comment]
Hi @0xmiroslav, could you take a look at the current proposals?

**Comment by ShogunFire** [comment]
@0xmiroslav Hello, thanks for reviewing my proposal. Can you give me more feedback about my proposal so that I can improve and maybe make a better proposal ? You don't like it because it uses AppState or because of something else ?

**Comment by marcochavezf** [comment]
Also, considering the posted proposals, what do you think would be the best approach to address this issue? Since the problem occurs when switching to another tab, I'm wondering if the solution should be limited to the web only, thoughts?

**Comment by michaelhaxhiu** [comment]
@0xmiroslav should this be on hold for https://github.com/Expensify/App/pull/18065? Let's update the GH title if so.

**Comment by dukenv0307** [comment]
@0xmiroslav sure thing

[Proposal updated](https://github.com/Expensify/App/issues/16576#issuecomment-1509730765)

**Comment by marcochavezf** [comment]
Not overdue, waiting for @0xmiroslav to review the updated proposal

**Comment by marcochavezf** [comment]
On hold for https://github.com/Expensify/App/issues/16364

**Comment by michaelhaxhiu** [comment]
still held 🕳️

**Comment by marcochavezf** [comment]
On hold

**Comment by marcochavezf** [comment]
On hold

**Comment by michaelhaxhiu** [comment]
same hold situation

**Comment by marcochavezf** [comment]
Still on hold

**Comment by michaelhaxhiu** [comment]
going `weekly` because melvin is getting on my bad side 😂  (also - "hold" in title is the req for `weekly` label)

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
> Still on hold

**Comment by marcochavezf** [comment]
> Still on hold

**Comment by marcochavezf** [comment]
> Still on hold

**Comment by marcochavezf** [comment]
Still on hold https://github.com/Expensify/App/issues/16364

**Comment by michaelhaxhiu** [comment]
Note: I'm preparing to go OOO for ~2 weeks, and need a buddy to oversee this while I'm gone.

@sonialiap in terms of next steps:
- This is on "hold" until https://github.com/Expensify/App/issues/16364 is finalized, since there's a dependency of code on the proposal here and what's being implemented in #16364.
- There's a good chance this doesn't see any movement while I'm gone. In which case I'll just take it over when I'm back!

**Comment by marcochavezf** [comment]
Since this is on hold for a weekly issue that's also on hold, I think it would make sense to change it to weekly

**Comment by sonialiap** [comment]
Moving to weekly

**Comment by michaelhaxhiu** [comment]
Back from OOO, unassigning Sonia.

**Comment by michaelhaxhiu** [comment]
two GHs are holding this one, and so I'm using `monthly` for now.

**Comment by michaelhaxhiu** [comment]
No progress on this still as it's blocked on. I'm going to manually re-assign to @sonialiap since she had an eye on it before. 

**Next steps** (same as before):

- This is on "hold" until https://github.com/Expensify/App/issues/16364 is finalized, since there's a dependency of code on the proposal here and what's being implemented there.
- If this isn't fixed by the time I return, I'm glad to take it back! fingers crossed 🤞 it gets unblocked fixed though

**Comment by sonialiap** [comment]
still on hold for https://github.com/Expensify/App/issues/16078

**Comment by sonialiap** [comment]
still on hold for https://github.com/Expensify/App/issues/16078

**Comment by mallenexpensify** [comment]
@DylanDylann reassigning, please take over as C+. If you don't have bandwidth, unassign yourself. Thanks

**Comment by DylanDylann** [comment]
@sonialiap Could you help to update the title: Hold on 16078 and update this issue to weekly because 16078 also be weekly

**Comment by marcochavezf** [comment]
Still on hold

**Comment by sonialiap** [comment]
still on hold for https://github.com/Expensify/App/issues/16078

**Comment by marcochavezf** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by marcochavezf** [comment]
> Still on hold for #16078

**Comment by sonialiap** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by DylanDylann** [comment]
still on hold

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by marcochavezf** [comment]
> Still on hold for #16078

**Comment by sonialiap** [comment]
Still on hold for https://github.com/Expensify/App/issues/16078

**Comment by DylanDylann** [comment]
Still on hold. @sonialiap Should we update label to Monthly

**Comment by marcochavezf** [comment]
Still on hold. I think weekly is fine, the other issue was moved to weekly too

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
The PR for the issue that we're holding for was recently [merged](https://github.com/Expensify/App/issues/16078), but let's wait a bit more in case of regressions

**Comment by sonialiap** [comment]
https://github.com/Expensify/App/issues/16078 has merged and the reported issue is reproducible

**Comment by sonialiap** [comment]
Asking team input on value of issue - [slack](https://expensify.slack.com/archives/C03U7DCU4/p1720002352779229)

**Comment by sonialiap** [comment]
After discussing we've decided to close this issue as low priority
